--- mkarchiso	2021-02-01 10:38:12.000000000 +0100
+++ mkarchiso	2021-02-04 22:16:33.402633314 +0100
@@ -294,6 +294,8 @@
     else
         pacstrap -C "${work_dir}/pacman.conf" -c -G -M -- "${airootfs_dir}" "${pkg_list[@]}"
     fi
+    sbsign --key DB.key --cert DB.crt --output "${airootfs_dir}/boot/vmlinuz-linux" "${airootfs_dir}/boot/vmlinuz-linux"
+    sbsign --key DB.key --cert DB.crt --output "${airootfs_dir}/usr/lib/systemd/boot/efi/systemd-bootx64.efi" "${airootfs_dir}/usr/lib/systemd/boot/efi/systemd-bootx64.efi"
 
     if [[ -n "${gpg_key}" ]]; then
         exec {ARCHISO_GNUPG_FD}<&-
@@ -476,6 +478,8 @@
     # the required image size in KiB (rounded up to the next full MiB with an additional MiB for reserved sectors)
     efiboot_imgsize="$(du -bc \
         "${airootfs_dir}/usr/lib/systemd/boot/efi/systemd-bootx64.efi" \
+        "${airootfs_dir}/usr/share/shim-signed/shimx64.efi" \
+        "${airootfs_dir}/usr/share/shim-signed/mmx64.efi" \
         "${airootfs_dir}/usr/share/edk2-shell/x64/Shell_Full.efi" \
         "${profile}/efiboot/" \
         "${airootfs_dir}/boot/vmlinuz-"* \
@@ -494,7 +498,11 @@
 
     mmd -i "${work_dir}/efiboot.img" ::/EFI ::/EFI/BOOT
     mcopy -i "${work_dir}/efiboot.img" \
-        "${airootfs_dir}/usr/lib/systemd/boot/efi/systemd-bootx64.efi" ::/EFI/BOOT/BOOTx64.EFI
+        "${airootfs_dir}/usr/lib/systemd/boot/efi/systemd-bootx64.efi" ::/EFI/BOOT/grubx64.efi
+    mcopy -i "${work_dir}/efiboot.img" \
+        "${airootfs_dir}/usr/share/shim-signed/mmx64.efi" ::/EFI/BOOT/mmx64.efi
+    mcopy -i "${work_dir}/efiboot.img" \
+        "${airootfs_dir}/usr/share/shim-signed/shimx64.efi" ::/EFI/BOOT/BOOTx64.EFI
 
     mmd -i "${work_dir}/efiboot.img" ::/loader ::/loader/entries
     mcopy -i "${work_dir}/efiboot.img" "${profile}/efiboot/loader/loader.conf" ::/loader/
